### PR TITLE
Trsh unit test

### PR DESCRIPTION
**Pull request by kspieks**
Added a unit test in the trshTest.py file that tests the trsh_ess_job function in trsh.py. I welcome your feedback!

### CODE REVIEW

File: arc/job/trsh.py [hunk 1]
```diff
@@ -381,7 +381,7 @@ def trsh_scan_job(label, scan_res, scan, species_scan_lists, methods):
 
 
 def trsh_ess_job(label, level_of_theory, server, job_status, job_type, software, fine, memory_gb,
-                 ess_trsh_methods=None, available_ess=None):
```
Comment (alongd) [inline arc/job/trsh.py]: :+1:

File: arc/job/trsh.py [hunk 2]
```diff
@@ -492,7 +492,9 @@ def trsh_ess_job(label, level_of_theory, server, job_status, job_type, software,
                 type=job_type, software=software, label=label))
             ess_trsh_methods.append('int=(Acc2E=14)')
             trsh_keyword = 'int=(Acc2E=14)'
-        elif 'cbs-qb3' not in ess_trsh_methods and level_of_theory != 'cbs-qb3' and 'scan' not in job_type:
+        # suggest spwaning a cbs-qb3 job if there are not many heavy atoms
+        elif 'cbs-qb3' not in ess_trsh_methods and level_of_theory != 'cbs-qb3' \
+                and 'scan' not in job_type and num_heavy_atoms<10:
```
Comment (alongd) [inline arc/job/trsh.py]: perhaps change to `<=`?

File: arc/job/trshTest.py [hunk 3]
```diff
@@ -97,5 +97,103 @@ def test_determine_ess_status(self):
         self.assertIn(' ? Basis library exhausted', line)  # line includes '\n'
 
 
+    def test_trsh_ess_job(self):
+        """Test the trsh_ess_job() function"""
+
+        #### test gaussian ####
+        label = 'ethanol'
+        level_of_theory = 'ccsd/vdz' # 'cbs-qb3'
+        server = 'rmg.mit.edu'
```
Comment (alongd) [inline arc/job/trshTest.py]: `server` should be s string, like `server1`

File: arc/job/trshTest.py [hunk 4]
```diff
@@ -97,5 +97,103 @@ def test_determine_ess_status(self):
         self.assertIn(' ? Basis library exhausted', line)  # line includes '\n'
 
 
+    def test_trsh_ess_job(self):
+        """Test the trsh_ess_job() function"""
+
+        #### test gaussian ####
+        label = 'ethanol'
+        level_of_theory = 'ccsd/vdz' # 'cbs-qb3'
```
Comment (alongd) [inline arc/job/trshTest.py]: Why the `cbs-qb3` comment?

File: arc/job/trshTest.py [hunk 5]
```diff
@@ -97,5 +97,103 @@ def test_determine_ess_status(self):
...
+        num_heavy_atoms = 2
+        ess_trsh_methods = ['change_node', 'int=(Acc2E=14)']
+
+        ## gaussian: test 1 ##
+        job_status = {'keywords': ['CheckFile']}
+        out = trsh.trsh_ess_job(label, level_of_theory, server, job_status, job_type,
+                                software, fine, memory_gb,num_heavy_atoms, ess_trsh_methods)
+        # unpack the output
+        output_errors, ess_trsh_methods, remove_checkfile, level_of_theory, software, job_type, fine, trsh_keyword,\
+        memory, shift, couldnt_trsh = out
```
Comment (alongd) [inline arc/job/trshTest.py]: perhaps unpack while calling `trsh_ess_job`?

File: arc/job/trshTest.py [hunk 6]
```diff
@@ -97,5 +97,103 @@ def test_determine_ess_status(self):
...
+
+        self.assertEqual(remove_checkfile, True)
```
Comment (alongd) [inline arc/job/trshTest.py]: Change to `self.assertTrue` (and below to `assertFalse`)

File: arc/job/trshTest.py [hunk 7]
```diff
@@ -97,5 +97,103 @@ def test_determine_ess_status(self):
...
+        out = trsh.trsh_ess_job(label, level_of_theory, server, job_status, job_type,
+                                software, fine, memory_gb, num_heavy_atoms, ess_trsh_methods)
+        # unpack the output
+        output_errors, ess_trsh_methods, remove_checkfile, level_of_theory, software, job_type, fine, trsh_keyword, \
+        memory, shift, couldnt_trsh = out
+
+        self.assertIn('cbs-qb3', ess_trsh_methods)
+        self.assertEqual(level_of_theory, 'cbs-qb3')
+        self.assertEqual(job_type, 'composite')
+
```
Comment (alongd) [inline arc/job/trshTest.py]: I see that between different ESS you kept 2 empty lines to make a distinction, but [PEP8](https://www.python.org/dev/peps/pep-0008/) asks that we keep one at most

File: arc/job/trshTest.py [hunk 8]
```diff
@@ -97,5 +97,103 @@ def test_determine_ess_status(self):
...
+        #### test molpro ####
+        software = 'molpro'
+
+        ## molpro: test ##
+        path = os.path.join(self.base_path['molpro'], 'insufficient_memory.out')
+        label = 'TS'
+        level_of_theory = 'mrci/aug-cc-pV(T+d)Z'
+        server = 'server1'
+        status, keywords, error, line = trsh.determine_ess_status(output_path=path, species_label='TS', job_type='sp')
+        job_status = {'keywords' : []}
```
Comment (alongd) [inline arc/job/trshTest.py]: perhaps instead you could do this as one line:
```python
job_status = {'keywords': keywords, 'error': error}
```

File: arc/job/trshTest.py [hunk 9]
```diff
@@ -97,5 +97,103 @@ def test_determine_ess_status(self):
...
+        fine = True
+        memory_gb = 32.0
+        ess_trsh_methods = ['change_node']
+        out = trsh.trsh_ess_job(label, level_of_theory, server, job_status, job_type,
+                                software, fine, memory_gb, num_heavy_atoms, ess_trsh_methods)
+        # unpack the output
+        output_errors, ess_trsh_methods, remove_checkfile, level_of_theory, software, job_type, fine, trsh_keyword,\
+        memory, shift, couldnt_trsh =out
+
+        self.assertIn('memory', ess_trsh_methods)
```
Comment (alongd) [inline arc/job/trshTest.py]: Could you also test the amount of memory this troubleshooting method suggests to use instead of 32 GB?

**Comment by alongd** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Clones added
============
- arc/job/trshTest.py  17
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/ARC/ARC/pullRequest?prid=4495605&bid=15158941)

File: arc/job/trshTest.py (hunk 7 above)
Comment (kspieks) [reply arc/job/trshTest.py]: Thanks! I'll check the guidelines next time :)

File: arc/job/trshTest.py (hunk 4 above)
Comment (kspieks) [reply arc/job/trshTest.py]: It's not needed.
I was trying different theories but needed a non 'cbs-qb3' theory to test the if statement that actually uses the `num_heavy_atoms`. I was sloppy and forgot to delete it so thanks for catching that!

File: arc/job/trshTest.py (hunk 9 above)
Comment (kspieks) [reply arc/job/trshTest.py]: Absolutely! The output seems to make sense to me. Here's the math. The out file had `insufficient memory available - require            23661817216  have 
            3999999059`
The function `determine_ess_status` converts the required memory to MW, resulting in 23661.817216. `trsh_ess_job` then rounds that up to the nearest 100, resulting in 23700. Dividing by 128 seems to be a conservative estimate? (https://deviceanalytics.com/memcalc/). Adding 5 GB is another good conservative measure and then the final step is to add the "additional required memory" to the original amount (32 GB in this case).

**Comment by alongd** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Clones added
============
- arc/job/trshTest.py  17
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/ARC/ARC/pullRequest?prid=4495605&bid=15158941)

**Comment by alongd** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Clones added
============
- arc/job/trshTest.py  12
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/ARC/ARC/pullRequest?prid=4495605&bid=15158941)

File: arc/job/trshTest.py (hunk 9 above)
Comment (alongd) [reply arc/job/trshTest.py]: Thanks for double-checking!

**Comment by kspieks** [comment]
> Looking great! Go ahead and squash the commits.

Squashed! Thanks for reviewing :)

**Comment by alongd** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Clones added
============
- arc/job/trshTest.py  12
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/ARC/ARC/pullRequest?prid=4495605&bid=15158941)

**Pull request merged**